File: solutions/1046.last-stone-weight.java
/*
 * @lc app=leetcode id=1046 lang=java
 *
 * [1046] Last Stone Weight
 */

// @lc code=start

import java.util.Comparator;
import java.util.PriorityQueue;

class Solution {
    public int lastStoneWeight(int[] stones) {
        PriorityQueue<Integer> queue = new PriorityQueue<>(stones.length, Comparator.reverseOrder());
        for (int stone : stones) {
            queue.add(stone);
        }

        while (queue.size() > 1) {
            int stone1 = queue.poll();
            int stone2 = queue.poll();
            if (stone1 != stone2) {
                queue.add(stone1 - stone2);
            }
        }

        return queue.isEmpty() ? 0 : queue.poll();
    }
}
// @lc code=end

File: solutions/1046.last-stone-weight.py
#
# @lc app=leetcode id=1046 lang=python3
#
# [1046] Last Stone Weight
#

# @lc code=start
import heapq
from typing import List


class Solution:
    def lastStoneWeight(self, stones: List[int]) -> int:
        heap_stones = [-s for s in stones]
        heapq.heapify(heap_stones)

        while len(heap_stones) > 1:
            stone1 = heapq.heappop(heap_stones)
            stone2 = heapq.heappop(heap_stones)
            if stone1 != stone2:
                heapq.heappush(heap_stones, stone1 - stone2)

        return -heap_stones[0] if heap_stones else 0

# @lc code=end

